tools/update-dll-msvc.bat

```diff
@@ -21,29 +18,106 @@ rem \     'unix' : 'gmake',
 rem \    },
 rem \ }
 
+setlocal enabledelayedexpansion
+
 if "%PROCESSOR_ARCHITECTURE%"=="AMD64" (
     set vimproc_arch=64
-    set cpu_arch=AMD64
-    set msvc_arch=amd64
 ) else (
     set vimproc_arch=32
-    set cpu_arch=x86
-    set msvc_arch=x86
 )
 
 set vimproc_dllname=vimproc_win%vimproc_arch%.dll
 
+if exist "%~dp0vswhere.exe" (
+    REM  found the lasted version of Visual C compiler
+    for /f "usebackq tokens=*" %%i in (`"%~dp0vswhere.exe" -latest -products * -requires Microsoft.VisualStudio.Component.VC.Tools.x86.x64 -property installationPath`) do (
+        set InstallDir=%%i
+
+        if exist "%InstallDir%\VC\Auxiliary\Build\Microsoft.VCToolsVersion.default.txt" (
+            set /p msvc_ver=<"%InstallDir%\VC\Auxiliary\Build\Microsoft.VCToolsVersion.default.txt"
+
+            rem Trim
+            set msvc_ver=!msvc_ver: =!
+        )
+
+        if exist "%InstallDir%\VC\Auxiliary\Build\vcvars%vimproc_arch%.bat" (
+            call "%InstallDir%\VC\Auxiliary\Build\vcvars%vimproc_arch%.bat"
+        )
+    )
+) else if defined VS140COMNTOOLS (
+    REM  Microsoft Visual Studio 2015 installed
+    set msvc_ver=14
+    call "%VS140COMNTOOLS%\..\..\VC\bin\vcvars%vimproc_arch%.bat"
+) else if defined VS120COMNTOOLS (
+    REM  Microsoft Visual Studio 2013 installed
+    set msvc_ver=12
+    call "%VS120COMNTOOLS%\..\..\VC\bin\vcvars%vimproc_arch%.bat"
+) else if defined VS110COMNTOOLS (
+    REM  Microsoft Visual Studio 2012 installed
+    set msvc_ver=11
+    call "%VS110COMNTOOLS%\..\..\VC\bin\vcvars%vimproc_arch%.bat"
+) else if defined VS100COMNTOOLS (
+    REM  Microsoft Visual Studio 2010 installed
+    set msvc_ver=10
+    call "%VS100COMNTOOLS%\..\..\VC\bin\vcvars%vimproc_arch%.bat"
+) else if defined VS90COMNTOOLS (
+    REM  Microsoft Visual Studo 2008 installed
+    set msvc_ver=9
+    call "%VS90COMNTOOLS%\..\..\VC\bin\vcvars%vimproc_arch%.bat"
+) else if defined VS80COMNTOOLS (
+    REM  Microsoft Visual Studio 2005 installed
+    set msvc_ver=8
+    call "%VS80COMNTOOLS%\..\..\VC\bin\vcvars%vimproc_arch%.bat"
+) else if defined VS71COMNTOOLS (
+    REM  Microsoft Visual C++ .NET 2003 installed
+    set msvc_ver=7
+    call "%VS71COMNTOOLS%\..\..\VC\bin\vcvars%vimproc_arch%.bat"
+) else if defined VS70COMNTOOLS (
+    REM  Microsoft Visual C++ .NET 2002 installed
+    set msvc_ver=7
+    call "%VS70COMNTOOLS%\..\..\VC\bin\vcvars%vimproc_arch%.bat"
+) else if exist "%ProgramFiles(x86)%\Microsoft Visual Studio 14.0\VC\bin\vcvars%vimproc_arch%.bat" (
```

Comment:
Do we need to check this?
Isn't it enough to check `%VSxxxCOMNTOOLS%`?

Reply:
@k-takata,  the design idea for update-dll-msvc.bat is if user has installed the Visual C compiler which version is 2017 or newer, we can use vswhere.exe with option '-latest', '-products' and '-requires' option.  vswhere.exe can detect the legacy version till to Visual Studio 10.0 but the user must install Visual C++ Build Tools for 2017 or newer at least or IDE version. So that previous I add the vswhere.exe in to git repositories. That is why except for to use vswhere.exe, I have also use VSxxxCOMNTOOLS environment variables to detect the location of Visual Studio installation path. For express edition of visual studio, the computer don't have VSxxxCOMNTOOLS environment variable. That is why I use the file path to detect the installation position. But this will be have a problem if the user don't install the Visual Studio Express edition not at %ProgramFiles(x86)%.

Reply:
> For express edition of visual studio, the computer don't have VSxxxCOMNTOOLS environment variable.

I install VS 2010, 2012, 2013 and 2015 (all Express), and all of them have VSxxxCOMNTOOLS variable.